### DISCUSSION TITLE
Given an E-class, is it possible to sample a batch of equivalent terms from it?

### DISCUSSION THREAD

**Discussion by brando90** [egraphs-good/egg#174] [Q&A]
I am interested to sample subset of terms from an e-class (in an E-graph). For example, in decreasing ast size/depth or even randomly. Is this possible? If yes how would one do it in egg?

**Comment by mwillsey** [comment]
Currently there isn't a built-in way to do it in egg. You can only extract one term from an e-class for a given cost function.

You could certainly build something to sample terms from an e-class by selecting random "representative' e-nodes for each e-class. Care must be taken to avoid cycles though. Here's [some code](https://github.com/egraphs-good/egg/blob/269c78636f91674da8c38838410d6fbdea08e099/src/lp_extract.rs#L206) that helps you locate e-nodes that make cycles.

**Reply by brando90** [reply]
so another option could be to select an e-node randomly (code I'd have to write in rust I assume) and feed it somehow to an e-class in an e-graph somehow?

Another hack is if there is a list of cost functions I could feed them one by one to get as many terms as that would allow me to get, is there such a list? I know ast-size and ast-depth exist. Any other?

**Reply by brando90** [reply]
@mwillsey max! what about this idea. Compute the length of the longest/biggest term. So the size of the term. Then feed as a "cost function" a value that returns infinity if the lengths don't match and 0 if they do match. Then this would at least return all terms of length L, L-1, L-2 if we change the cost function appropriately by changing the check for equality of size.

No sure how yet how to extract terms of equal length...also this would be O(N*k) for the times we call extract. Ideally there is some sort of dynamic programming that could be done when extracting say terms of size L to a smaller value...

**Reply by brando90** [reply]
actually let's take a step back to your idea of selecting e-nodes. Given an e-graph, how do I know which one is the "root e-class" (since the e-graph really is a set of asts so one must have been our starting node). Then I am find just selecting a random set of e-nodes (at least at first). 

Later it would be nice to do what @bastacyclop did and extract the N best terms according to a cost function.

**Reply by mwillsey** [reply]
The root e-class is a matter of perspective. You might have 0, 1, or many roots depending on what you inserted into the initial e-graph. Typically though, the root e-class is the e-class containing the initial term you added to the e-graph.

@Bastacyclop's Scala version is definitely a good source of inspiration, but if you just need to get an idea of what's in an e-class via sampling, I think the idea above would do just fine.

**Reply by brando90** [reply]
Thanks max! Just to clarify, what do you mean by "idea above"? Do you mean the one I suggested with a cost function that checks for `size(ast) == number` and returns 0 or 1 or something else?

Thanks again! :) Love egg and e-graphs.

**Reply by mwillsey** [reply]
[This one](https://github.com/egraphs-good/egg/discussions/174#discussioncomment-2433405): just selecting random e-nodes after cycle removal. Glad you're having fun!

**Reply by sqPoseidon** [reply]
@brando90 Hi, do you have any updates on this problem? I want to sample the equivalent terms too. Many thanks if you have any comments or suggestions.

**Comment by Bastacyclop** [comment]
I've personally implemented in a Scala port:
- [extracting the N best terms of an e-class according to a cost function](https://github.com/rise-lang/shine/blob/6a44d8c10b8eb8f4e05b8b076964588f20bdd3c4/src/main/scala/rise/eqsat/Analysis.scala#L506-L558)
- [counting programs up to a certain size](https://github.com/rise-lang/shine/blob/main/src/main/scala/rise/eqsat/Analysis.scala#L945-L981)
- [extracting random programs](https://github.com/rise-lang/shine/blob/main/src/main/scala/rise/eqsat/Extractor.scala#L12-L22), however this recursively makes a random choice of e-node so the random distribution is not great (cycles may also be taken but I assume that the probability of exiting a cycle is high). I think that the random distribution could be improved by combining this with counting and a size limit (e.g. weigh the e-node choice by how many programs it represents).

**Reply by Bastacyclop** [reply]
@mwillsey Any idea how extracting the N best terms would work with LP extraction? Could we duplicate the variables to solve N times to get N solutions, and include the constraint that each solution should be different?

**Reply by mwillsey** [reply]
Hmm, interesting. I think it would be possible to do that, but I think you'll have a problem with diversity because of equivalent subterms. For example, f(g(a + b)) = f(g(b + a)). For "top k" extraction, I think my best idea so far is extract, delete a random e-node from the e-graph that was in the extracted term, and repeat.

**Reply by Bastacyclop** [reply]
I feel like the diversity issue would benefit from a problem-specific (dis)similarity function, see:
https://link.springer.com/chapter/10.1007/978-3-642-33715-4_1
https://arxiv.org/abs/1610.02424
https://dl.acm.org/doi/abs/10.1145/3485486

Maybe its possible to include such a function into the LP objective (i.e. maximize diversity between the N solutions), but that might be difficult / costly.

**Comment by brando90** [comment]
I don't unfortunately :(

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by sqPoseidon** [reply]
Emm, thank you. 
@mwillsey Hi Max, you've mentioned that we can delete some enodes from the egraph. Can you provide me with some examples of how to delete the nodes? I didn't find that in the document.

**Reply by mwillsey** [reply]
You can reach into [`EClass::nodes`](https://docs.rs/egg/latest/egg/struct.EClass.html#structfield.nodes) and do whatever you want! It's just a Vec of Enodes.
